File: frontend/src/lib/combatReducer.ts
// TypeScript twin of backend/app/modules/playbook/combat_reducer.py (ADR-005). Powers the
// combat tracker's optimistic UI; kept byte-for-byte identical via shared golden fixtures.
//
// Nothing here rolls a die, and nothing here may — folding the log has to be deterministic
// or undo/redo silently corrupts the combat. Rolls resolve server-side and arrive as literal
// results: set_initiative {value: 17}, never roll_initiative {}. Any change here must land in
// the Python reference and backend/tests/fixtures/combat_golden.json in the same commit.

export interface Combatant {
  id: string
  name: string
  side: string
  /** 'creature' | 'lair'. A lair sits in the initiative order as an ordinary entry. */
  kind: string
  max_hp: number
  hp: number
  temp_hp: number
  initiative: number
  /** Breaks ties before falling back to id (5e: the dex modifier). */
  initiative_tiebreak: number
  conditions: string[]
  concentrating: boolean
  defeated: boolean
  death_saves: { successes: number; failures: number }
  legendary: { max: number; remaining: number }
}

export interface CombatState {
  round: number
  turn_index: number
  order: string[]
  combatants: Record<string, Combatant>
}

export type Action = { type: string } & Record<string, unknown>

// Match Python's int(): coerce to a number, then truncate toward zero. HP/initiative are
// integers, and the reference reducer applies int() to every numeric field — so the twin
// must truncate identically or the two folds diverge on a fractional amount (ADR-005).
const toInt = (x: unknown): number => Math.trunc(Number(x))

export function initialState(): CombatState {
  return { round: 1, turn_index: 0, order: [], combatants: {} }
}

function reorder(state: CombatState): void {
  const ids = Object.keys(state.combatants)
  ids.sort((a, b) => {
    const ca = state.combatants[a]
    const cb = state.combatants[b]
    const d = cb.initiative - ca.initiative
    if (d !== 0) return d
    const t = cb.initiative_tiebreak - ca.initiative_tiebreak
    if (t !== 0) return t
    return a < b ? -1 : a > b ? 1 : 0
  })
  state.order = ids
  state.turn_index = ids.length ? Math.min(state.turn_index, ids.length - 1) : 0
}

// A lair at 0 hp is not a corpse — it has no hit points to lose in the first place.
function isDefeated(m: Combatant): boolean {
  return m.hp === 0 && m.kind !== 'lair'
}

export function applyAction(state: CombatState, action: Action): CombatState {
  const kind = action.type
  const c = state.combatants
  const id = action.id as string

  if (kind === 'add_combatant') {
    const maxHp = toInt(action.max_hp)
    const hp = action.hp === undefined ? maxHp : toInt(action.hp)
    const entryKind = (action.kind as string) ?? 'creature'
    const legendaryMax = toInt(action.legendary_max ?? 0)
    c[id] = {
      id, name: String(action.name), side: (action.side as string) ?? 'foe',
      kind: entryKind,
      max_hp: maxHp, hp, temp_hp: 0, initiative: toInt(action.initiative ?? 0),
      initiative_tiebreak: toInt(action.initiative_tiebreak ?? 0),
      conditions: [], concentrating: false,
      defeated: hp <= 0 && entryKind !== 'lair',
      death_saves: { successes: 0, failures: 0 },
      legendary: { max: legendaryMax, remaining: legendaryMax },
    }
    reorder(state)
  } else if (kind === 'set_initiative') {
    if (c[id]) {
      c[id].initiative = toInt(action.value)
      // Rolling sends the tiebreak along; a manual edit from the rail omits it and leaves
      // whatever the combatant was seeded with.
      if (action.initiative_tiebreak !== undefined) {
        c[id].initiative_tiebreak = toInt(action.initiative_tiebreak)
      }
      reorder(state)
    }
  } else if (kind === 'damage') {
    const m = c[id]
    if (m) {
      const alreadyDown = m.hp === 0 && m.kind !== 'lair'
      let amount = toInt(action.amount)
      const absorbed = Math.min(m.temp_hp, amount)
      m.temp_hp -= absorbed
      amount -= absorbed
      m.hp = Math.max(0, m.hp - amount)
      m.defeated = isDefeated(m)
      if (m.hp === 0) m.concentrating = false
      // Hitting someone who is already down is an automatic failed death save — the single
      // most-forgotten rule at a table, and free to get right here.
      if (alreadyDown && amount > 0) m.death_saves.failures += 1
    }
  } else if (kind === 'heal') {
    const m = c[id]
    if (m) {
      m.hp = Math.min(m.max_hp, m.hp + toInt(action.amount))
      if (m.hp > 0) {
        m.defeated = false
        // Back above 0: nobody is dying any more, so the clock resets.
        m.death_saves = { successes: 0, failures: 0 }
      }
    }
  } else if (kind === 'death_save') {
    const m = c[id]
    if (m) {
      // The *outcome* was decided server-side (a natural 20 is the plugin's rule, not this
      // module's); all that happens here is bookkeeping on a literal result.
      const result = action.result as string
      if (result === 'crit_success') {
        m.hp = Math.min(m.max_hp, 1)
        m.defeated = isDefeated(m)
        m.death_saves = { successes: 0, failures: 0 }
      } else if (result === 'crit_fail') {
        m.death_saves.failures += 2
      } else if (result === 'success') {
        m.death_saves.successes += 1
      } else if (result === 'failure') {
        m.death_saves.failures += 1
      }
    }
  } else if (kind === 'set_temp_hp') {
    if (c[id]) c[id].temp_hp = Math.max(0, toInt(action.amount))
  } else if (kind === 'add_condition') {
    const m = c[id]
    const cond = String(action.condition)
    if (m && !m.conditions.includes(cond)) m.conditions.push(cond)
  } else if (kind === 'remove_condition') {
    const m = c[id]
    const cond = String(action.condition)
    if (m) m.conditions = m.conditions.filter((x) => x !== cond)
  } else if (kind === 'set_concentration') {
    if (c[id]) c[id].concentrating = Boolean(action.on)
  } else if (kind === 'next_turn') {
    if (state.order.length) {
      state.turn_index += 1
      if (state.turn_index >= state.order.length) {
        state.turn_index = 0
        state.round += 1
      }
    }
  } else if (kind === 'remove_combatant') {
    if (c[id]) { delete c[id]; reorder(state) }
  }
  return state
}

export function fold(actions: Action[]): CombatState {
  const state = initialState()
  for (const a of actions) applyAction(state, a)
  return state
}

// Apply one action to a copy (for optimistic UI).
export function applyOptimistic(state: CombatState, action: Action): CombatState {
  const copy: CombatState = JSON.parse(JSON.stringify(state))
  return applyAction(copy, action)
}

File: backend/app/modules/playbook/combat_reducer.py
"""Pure combat-state reducer (ADR-005): state = fold(actions).

Combat is event-sourced *inside* an encounter run: the state is never stored per-field,
only derived by folding the action log. Undo/redo is a cursor into that log. This module
is a system-agnostic reducer (HP, conditions, turn order as data); it has a byte-for-byte
TypeScript twin (frontend/src/lib/combatReducer.ts) verified against shared golden fixtures.

**Nothing here rolls a die, and nothing here may.** Folding the log must be deterministic —
replaying it has to land on the same state every time, or undo/redo silently corrupts the
combat. Rolls resolve server-side (``app.core.dice``) and reach this module only as literal
results: ``set_initiative {value: 17}``, never ``roll_initiative {}``.

Payload keys the reducer doesn't read are legal and ignored (``stat_block_id``, ``roll_id``),
which is how provenance rides along without touching the folded state or its fixtures.
"""

from __future__ import annotations

from typing import Any, Literal, get_args

Action = dict[str, Any]
State = dict[str, Any]

#: The complete action vocabulary — every type ``apply_action`` below actually handles.
#: It lives here, next to the implementation, so the API's validation cannot drift from what
#: the reducer understands. Add a type here and to ``apply_action`` in the same change; an
#: unhandled name would otherwise persist to the log, advance the cursor, and do nothing.
ActionType = Literal[
    "add_combatant",
    "remove_combatant",
    "set_initiative",
    "damage",
    "heal",
    "set_temp_hp",
    "add_condition",
    "remove_condition",
    "set_concentration",
    "death_save",
    "legendary_use",
    "cast_spell",
    "next_turn",
]
ACTION_TYPES: tuple[str, ...] = get_args(ActionType)


def initial_state() -> State:
    return {"round": 1, "turn_index": 0, "order": [], "combatants": {}}


def _reorder(state: State) -> None:
    combatants = state["combatants"]
    # Ties break on the tiebreak (5e: the dex modifier), then on id. Before initiative was
    # ever rolled the tiebreak was moot; with real d20s, ties are common and breaking them
    # on an opaque id would be arbitrary.
    ids = sorted(
        combatants,
        key=lambda i: (
            -combatants[i]["initiative"],
            -combatants[i]["initiative_tiebreak"],
            i,
        ),
    )
    state["order"] = ids
    state["turn_index"] = min(state["turn_index"], len(ids) - 1) if ids else 0


def _defeated(combatant: dict[str, Any]) -> bool:
    """A lair at 0 hp is not a corpse — it has no hit points to lose in the first place."""
    return bool(combatant["hp"] == 0 and combatant["kind"] != "lair")


def apply_action(state: State, action: Action) -> State:
    kind = action["type"]
    combatants: dict[str, Any] = state["combatants"]

    if kind == "add_combatant":
        cid = action["id"]
        max_hp = int(action["max_hp"])
        hp = int(action.get("hp", max_hp))
        entry_kind = action.get("kind", "creature")
        legendary_max = int(action.get("legendary_max", 0))
        combatants[cid] = {
            "id": cid, "name": action["name"], "side": action.get("side", "foe"),
            # "creature" | "lair". A lair rides in the initiative order as an ordinary
            # entry (5e: count 20) so its turn needs no special case anywhere.
            "kind": entry_kind,
            "max_hp": max_hp, "hp": hp, "temp_hp": 0,
            "initiative": int(action.get("initiative", 0)),
            "initiative_tiebreak": int(action.get("initiative_tiebreak", 0)),
            "conditions": [], "concentrating": False,
            "defeated": hp <= 0 and entry_kind != "lair",
            "death_saves": {"successes": 0, "failures": 0},
            "legendary": {"max": legendary_max, "remaining": legendary_max},
            # {pool key: {"label", "level", "max", "remaining"}} — spell slots and innate
            # per-day uses, seeded by the rule system. Empty for anything that can't cast.
            # The keys are the plugin's; nothing in here reads inside them.
            "spell_pools": dict(action.get("spell_pools") or {}),
        }
        _reorder(state)
    elif kind == "set_initiative":
        m = combatants.get(action["id"])
        if m:
            m["initiative"] = int(action["value"])
            # Rolling initiative sends the tiebreak along with the value; a manual edit
            # from the rail omits it and leaves whatever the combatant was seeded with.
            if "initiative_tiebreak" in action:
                m["initiative_tiebreak"] = int(action["initiative_tiebreak"])
            _reorder(state)
    elif kind == "damage":
        m = combatants.get(action["id"])
        if m:
            already_down = m["hp"] == 0 and m["kind"] != "lair"
            amount = int(action["amount"])
            absorbed = min(m["temp_hp"], amount)
            m["temp_hp"] -= absorbed
            amount -= absorbed
            m["hp"] = max(0, m["hp"] - amount)
            m["defeated"] = _defeated(m)
            if m["hp"] == 0:
                m["concentrating"] = False
            # Hitting someone who is already down is an automatic failed death save — the
            # single most-forgotten rule at a table, and free to get right here.
            if already_down and amount > 0:
                m["death_saves"]["failures"] += 1
    elif kind == "heal":
        m = combatants.get(action["id"])
        if m:
            m["hp"] = min(m["max_hp"], m["hp"] + int(action["amount"]))
            if m["hp"] > 0:
                m["defeated"] = False
                # Back above 0: nobody is dying any more, so the clock resets.
                m["death_saves"] = {"successes": 0, "failures": 0}
    elif kind == "death_save":
        m = combatants.get(action["id"])
        if m:
            # The *outcome* was decided server-side (a natural 20 is the plugin's rule, not
            # this module's); all that happens here is bookkeeping on a literal result.
            result = action["result"]
            saves = m["death_saves"]
            if result == "crit_success":
                m["hp"] = min(m["max_hp"], 1)
                m["defeated"] = _defeated(m)
                m["death_saves"] = {"successes": 0, "failures": 0}
            elif result == "crit_fail":
                saves["failures"] += 2
            elif result == "success":
                saves["successes"] += 1
            elif result == "failure":
                saves["failures"] += 1
    elif kind == "set_temp_hp":
        m = combatants.get(action["id"])
        if m:
            m["temp_hp"] = max(0, int(action["amount"]))
    elif kind == "add_condition":
        m = combatants.get(action["id"])
        if m and action["condition"] not in m["conditions"]:
            m["conditions"].append(action["condition"])
    elif kind == "remove_condition":
        m = combatants.get(action["id"])
        if m and action["condition"] in m["conditions"]:
            m["conditions"].remove(action["condition"])
    elif kind == "set_concentration":
        m = combatants.get(action["id"])
        if m:
            m["concentrating"] = bool(action["on"])
    elif kind == "legendary_use":
        m = combatants.get(action["id"])
        if m:
            cost = int(action.get("cost", 1))
            m["legendary"]["remaining"] = max(0, m["legendary"]["remaining"] - cost)
    elif kind == "cast_spell":
        m = combatants.get(action["id"])
        # Unlike legendary actions, a spent slot never comes back on a turn: a monster's
        # pools refill because the next fight seeds fresh ones, and a character's because
        # they rested. An unknown pool is ignored — a cantrip has none, and the tracker
        # doesn't log those at all.
        if m:
            pool = m["spell_pools"].get(action.get("pool_key"))
            if pool:
                pool["remaining"] = max(0, pool["remaining"] - int(action.get("cost", 1)))
    elif kind == "next_turn":
        if state["order"]:
            state["turn_index"] += 1
            if state["turn_index"] >= len(state["order"]):
                state["turn_index"] = 0
                state["round"] += 1
            # A creature regains its spent legendary actions at the *start of its turn*
            # (5e), which is exactly now — so the reset rides the turn rather than needing
            # anyone to remember it.
            current = combatants.get(state["order"][state["turn_index"]])
            if current and current["legendary"]["max"] > 0:
                current["legendary"]["remaining"] = current["legendary"]["max"]
    elif kind == "remove_combatant":
        cid = action["id"]
        if cid in combatants:
            del combatants[cid]
            _reorder(state)

    return state


def fold(actions: list[Action]) -> State:
    state = initial_state()
    for action in actions:
        apply_action(state, action)
    return state
